modules/apps/transfer/keeper/forwarding.go

```diff
@@ -49,7 +49,11 @@ func (k Keeper) acknowledgeForwardedPacket(ctx sdk.Context, packet channeltypes.
 		return errorsmod.Wrap(channeltypes.ErrChannelCapabilityNotFound, "module does not own channel capability")
 	}
 
-	return k.ics4Wrapper.WriteAcknowledgement(ctx, capability, packet, ack)
+	if err := k.ics4Wrapper.WriteAcknowledgement(ctx, capability, packet, ack); err != nil {
+		return err
+	}
+
+	return k.deleteForwardedPacket(ctx, packet.SourcePort, packet.SourceChannel, packet.Sequence)
```

Comment:
```suggestion
	return k.deleteForwardedPacket(ctx, packet.DestinationPort, packet.DestinationChannel, packet.Sequence)
```

This confused me during the walkthrough a bit. During a middlehop, we forward a new packet. In storage, we make a mapping from the portID, channelID we send that packet from to the packet which caused its creation (creating a pointer in the reverse direction)

Thus when processing the ack/timeout on a middlehop, core IBC gives us the packet we created as the result of a forward. We use that source port/channel id to get the reverse direction packet. Thus when we delete the mapping to the reverse direction packet (`packet` arg in this function is `prevPacket` (maybe should be renamed)) we should use the destination port and channel as that matches the source port and channel used in the original set

Reply:
lets add an in-line code comment to explain this. I'm sure it'll take many code readers a second to parse

Reply:
> we should use the destination port and channel as that matches the source port and channel used in the original set (wording here was a bit long winded, we can maybe fine a more concise way to explain)

Sorry if I am missing something, but as @colin-axner says the `packet` used here is `prevPacket` (the packet that we receive in the middle hop) and this packet [is stored using the source port ID and channel ID of the next packet](https://github.com/cosmos/ibc-go/blob/c546f69b7ad0e312fe0102c172462b3669a7e650/modules/apps/transfer/keeper/relay.go#L283) (the packet we send from the middle chain), right?

And the destination port ID/channel ID of `prevPacket` does not necessarily match the source port ID/channel ID of the next packet, right? So shouldn't we be using here instead the source port ID/channel ID of the packet that was sent from the middle chain?

Again apologies if I am getting this wrong.

Reply:
Sorry I was incorrect in my above explanation! The packet received in a middlehop has the source port/channel from the chain that sent the packet and the dest port/channel which is connected to that sender chain. We then send a new packet on the first forwarding hop which is distinctly different from the destination port/channel

If we try to use `prevPacket` we need to unmarshal the data. I would argue we should not be storing the entire packet, but rather just the packetID info. Therefore, I would recommend instead, we break the existing `acknowledgeForwardSuccess` and `acknowledgePacketForwardError` to take in the packetID of the forwarded packet so you can use the source port/channel from that. Makes the api less clean, but I don't see another performant approach at the moment

Reply:
@crodriguezvega you are correct. It looks like we were writing our comments in parallel

Reply:
@colin-axner just talking with @bznein  now and trying to clarify exactly what needs to be done here.

Using as an example a three chain setup of A -> B -> C and ack is happening on B.

In this case, the forwarded packet is keyd under chain C's port/channel ID. And so we can get this either from

- the packet sent from A -> B's forwarding hops (from the packet data) [here](https://github.com/cosmos/ibc-go/blob/f9c1e70cab870be0f0736f848ee483cd257abdc0/modules/apps/transfer/keeper/forwarding.go#L47) 
 or
- the source port/channel ID in the packet received in the ack fn [here](https://github.com/cosmos/ibc-go/blob/60a6b9945beaf03820f6490c7853304c40601b2e/modules/apps/transfer/keeper/relay.go#L282).

Your suggestion is to not store the full packet, but it looks like we still need the `prevPacket` in some places like [here](https://github.com/cosmos/ibc-go/blob/f9c1e70cab870be0f0736f848ee483cd257abdc0/modules/apps/transfer/keeper/forwarding.go#L47)

Don't we still need to store the full packet for this to work? (technically we don't need the packet data, just the ports/channel IDS, is this what you meant by `packetID` info?)

Reply:
Thanks for the followup @chatton. See [packetID](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/types/packet.go#L132) type

>  the forwarded packet is keyd under chain C's port/channel ID.

yes, but more specifically the channel endpoint on chain B which connects to chain C

> but it looks like we still need the prevPacket in some places like [here](https://github.com/cosmos/ibc-go/blob/f9c1e70cab870be0f0736f848ee483cd257abdc0/modules/apps/transfer/keeper/forwarding.go#L47)

My recommendation is to only store the `packetID` https://github.com/cosmos/ibc-go/issues/6636 on the set you linked

> Don't we still need to store the full packet for this to work?

No. Only the packetID is necessary. We aren't acting on the previous data (we know the previous data since we constructed a packet from it). Once https://github.com/cosmos/ibc-go/issues/6635 is addressed, we won't need the full packet stored

See spec changes https://github.com/cosmos/ibc/pull/1119, I think our code abstractions actually made this a bit more complicated than it needs to be

Reply:
You can pass the full packet through no problem. I am just leaning towards the changes which would be implemented in https://github.com/cosmos/ibc-go/issues/6635, but happy to change things at that time period

Reply:
Thanks for the explanation Colin! 

I went with the solution of passing the packet just to be able to land this PR. I can happily take #6635 after that, but it would require me a bit to understand :)

Reply:
Thanks for the explanation @colin-axner and @bznein for pushing it forward :muscle: